### ISSUE TITLE
Feature proposal: better options management

### ISSUE THREAD

**Issue by darkgl0w** [fastify/fastify-jwt#27]
## 🚀 Feature Proposal

A way to manage `fastify-jwt` "global" default options.
Since v0.5.0 an `options` key has been introduced, this `options` key allows users to define some global options to use with `fastify-jwt` and thus introduced some unwanted side effects due to the way it has been implemented.

With this proposal I would like to introduce an `options` key like this:
```js
{
  options: {
    decode?: {
      complete?: boolean,
      json?: bool
    },
    sign?: {
      algorithm?: string,
      keyid?: string,
      expiresIn?: string | number,
      notBefore?: string | number,
      audience?: string | string[],
      subject?: string,
      issuer?: string,
      jwtid?: string,
      noTimestamp?: boolean,
      header?: object,
      encoding?: string
    },
    verify?: {
      algorithms?: string[],
      audience?: string | string[],
      clockTimestamp?: number,
      clockTolerance?: number,
      issuer?: string | string[],
      ignoreExpiration?: boolean,
      ignoreNotBefore?: boolean,
      jwtid?: string,
      subject?: string,
      maxAge?: string // Deprecated but still accepted by jsonwebtoken package
    }
  }
}
```
In the meantime I would like to introduce a new decoration named `options` to give users an easy way to access those default global options by calling `fastify.jwt.options` or more specifically `fastify.jwt.options.decode`, `fastify.jwt.options.sign` or `fastify.jwt.options.verify`(given that they have been defined or not).

And those global default options can be overridden by providing 'local' options when the user needs to (check the examples below)

NB: To avoid any breaking change with the `v0.5.0` that introduced an `options` key that took the shape `options: { algorithm?: string }` (this is the shape that is explicitly documented and thus the one susceptible to be already adopted by the users) I can take into account an extra key `algorithm?: string` inside of the options and then merge it inside the `sign` default options. We would need to mark it as deprecated and encourage users to adopt the shape that come with this proposal and then remove that deprecated practice in the version `0.8.0`  of the package (if we consider this feature is good to land with the `0.7.0` version of the package). To be honest this is not really needed as I don't think there is too much applications that are concerned, but still it's better to do it this away and let users the time to change their code.

## Motivation

Currently managing "global" default options within the package can be a bit confusing, because there is not really any actual filtering on the different accepted options and no real means to define global options for the different methods of the package.
This proposal aims to make the use of `fastify-jwt` more intuitive by introducing a global options management within the package.

What do you think of those changes ? They could finalize the `fastify-jwt` wrapping of `jsonwebtoken` in an almost perfect way.

I have a PR that's almost ready (I already refactored the code and made it more easy to read alongside this feature implementation), all it needs to be complete now is a good documentation rework and a little refactoring of the tests to make them more readable (and i need to add a few more to achieve a better coverage).

I would also like to add a typescript definition file for the package (it can be a good way to avoid a misuse of the package) but I would definitely need some help for this (I mostly use ES only and I am not particularly familiar with Typescript, maybe there is an example typescript definition file for a fastify module that's considered very clean and well implemented - the one I found in the `fastify-mongodb` plugin repository should be my model maybe ?)

## Example
```js
'use strict'

require('make-promises-safe')
const { readFileSync } = require('fs')
const fastify = require('fastify')({
  logger: true
})

fastify.register(require('fastify-jwt'), {
  secret: {
    private: readFileSync('path/to/private/cert.pem'),
    public: readFileSync('path/to/public/cert.pem')
  },
  options: {
    decode: {
      complete: true
    },
    sign: {
      issuer: 'foo'
    },
    verify: {
      issuer: 'foo'
    },
    algorithm: 'RS256' // Deprecated but still valid to avoid any break in the code
  }
})

fastify.get('/', async (request, reply) => {
  try {
    // We use the default global options to generate our token
    const token = await reply.jwtSign({ foo: 'bar' })

    // return a token corresponding to the payload { foo: 'bar', iss: 'foo' }
    return { token }
  } catch (error) {
    reply.code(500)
    return { error }
  }
})

fastify.get('/verify', async (request, reply) => {
  try {
    // We use the default global options to verify the Auth bearer token
    const payload = await request.jwtVerify()

    // { payload: { "foo": "bar", "iat": 1540294345, "iss": "foo" } }
    return { payload }
  } catch (error) {
    reply.code(401)
    return { error }
  }
})

fastify.get('/signAlt', async (request, reply) => {
  try {
    // We clone the global sign options
    // localSignOptions = { algorithm: 'RS256', issuer: 'foo' }
    let localSignOptions = Object.assign({}, fastify.jwt.options.sign )

    // We change the global sign options as we need something
    // that's a little bit different
    // localSignOptions = { algorithm: 'RS256', issuer: 'bar' }
    localSignOptions.issuer = 'bar'
    // localSignOptions = { algorithm: 'RS256', issuer: 'bar', subject: 'test' }
    localSignOptions.subject = 'test'

    // We clone the global verify options
    // localVerifyOptions = { issuer: 'foo' }
    let localVerifyOptions = Object.assign({}, fastify.jwt.options.verify )

    // We change the global sign options as we need something that's a little bit different
    // localVerifyOptions = { issuer: 'bar' }
    localVerifyOptions.issuer = 'bar'
    // localVerifyOptions = { issuer: 'bar', subject: 'test' }
    localVerifyOptions.subject = 'test'

    // We override global signing options with the ones we defined
    // this will generate a token corresponding to the payload { foo: 'bar', iss: 'bar', sub: 'test' }
    const token = await reply.jwtSign({ foo: 'bar' }, localSignOptions)

    // We override our global verification options to match our local signing
    // options and thus avoid a jwt validation error
    const payload = await fastify.jwt.verify(token, localVerifyOptions)

    // { "payload": { "foo": "bar", "iat": 1540305337, "iss": "bar","sub": "test" } }
    return { payload }
  } catch (error) {
    reply.code(401)
    return { error }
  }
})

fastify.get('/decode', async (request, reply) => {
  try {
    // We generate a token using the global default signing options
    const token = await fastify.jwt.sign({ foo: 'bar' })

    // We use the global default decoding options
    const fullyDecodedToken = await fastify.jwt.decode(token)

    // We override the global decoding options
    const noComplete = await fastify.jwt.decode(token, { complete: false })

    /*
    {
      "fullyDecodedToken": {
        "header": {
          "alg": "RS256",
          "typ": "JWT"
        },
        "payload": {
          "foo": "bar",
          "iat": 1540305600,
          "iss": "foo"
        },
        "signature": "gVf5bzROYB4nPgQC0nbJTWCiJ3Ya51cyuP-N50cidYo"
      },
      "noComplete": {
        "foo": "bar",
        "iat": 1540305600,
        "iss": "foo"
      }
    }
    */
    return { fullyDecodedToken, noComplete }
  } catch (error) {
    reply.code(500)
    return { error }
  }
})

fastify.listen(1337, (err) => {
  if (err) throw err
})
```

**Comment by mcollina** [comment]
I'd recommend to follow https://www.fastify.io/docs/latest/TypeScript/#authoring-plugin-types for writing plugin types.

+1 on the proposal. I would actually drop the `options` prefix and do:


```js

fastify.register(require('fastify-jwt'), {
  secret: {
    private: readFileSync('path/to/private/cert.pem'),
    public: readFileSync('path/to/public/cert.pem')
  },
  decode: {
    complete: true
  },
  sign: {
    issuer: 'foo'
  },
  verify: {
    issuer: 'foo'
  }
})
```

Anyway, go for it!

I would also like to recommend we document all options.

**Comment by darkgl0w** [comment]
OK, that's great :)
I will go for it and drop the `options` prefix (I already have a working POC) as suggested. Depending on my free time I might be able to submit a PR soon (maybe tomorrow or after tomorrow).

Concerning the documentation I plan to do a little rework to make it more explicit about all the new stuff that has been added and the way to use it.

**Comment by mcollina** [comment]
Implemented in #28.
